code/datums/gamemode/role/time_agent.dm

```diff
@@ -0,0 +1,165 @@
+			Locate certain time-sensitive objects scattered around the station, so as to locate the time anomaly and use it for extraction.<br>\
+			This may not be the first time you visit this timeline, and it may not be the last.</span>")
+
+	to_chat(antag.current, "<span class='danger'>Remember that the items you are provided with are largely non-expendable. Try not to lose them, especially the jump charge, as it is your ticket home.</span>")
+
+
+/datum/role/time_agent/ForgeObjectives()
+	AppendObjective(/datum/objective/target/locate/random)
+	if(prob(30))
+		AppendObjective(/datum/objective/target/assassinate)
+	AppendObjective(/datum/objective/freeform/aid)
+
+
+/datum/role/time_agent/process()
+	var/list/datum/objective/jecties = objectives.GetObjectives()
+	if(!jecties.len || locate(/datum/objective/time_agent_extract) in objectives.GetObjectives())
+		return //Not set up yet
+	var/finished = TRUE
+	for(var/datum/objective/O in objectives.GetObjectives())
+		if(!O.IsFulfilled())
+			finished = FALSE
+			break
+	if(finished)
+		to_chat(antag.current, "<span class = 'notice'>Objectives complete. Triangulating anomaly location.</span>")
+		AppendObjective(/datum/objective/time_agent_extract)
+
+/datum/role/time_agent/OnPostSetup()
+	.=..()
+	if(istype(antag.current, /mob/living/carbon/human))
+		equip_time_agent(antag.current)
+		spawn_rand_maintenance(antag.current)
+		spawn()
+			showrift(antag.current,1)
+
+
+/datum/role/time_agent/proc/extract()
+	var/mob/living/carbon/human/H = antag.current
+	H.drop_all()
+	showrift(H,1)
+	qdel(H)
+	for(var/i in objects_to_delete)
+		objects_to_delete.Remove(i)
+		qdel(i)
+	increment_threat(rand(5,10))
+
+/obj/item/device/chronocapture
+	name = "chronocapture device"
+	desc = "Used to confirm that everything is where it should be."
+	icon = 'icons/obj/items.dmi'
+	icon_state = "polaroid"
+	item_state = "polaroid"
+	w_class = W_CLASS_SMALL
+	var/triggered = FALSE
+
+/obj/item/device/chronocapture/afterattack(atom/target, mob/user)
+	if(!triggered)
+		triggered = TRUE
+		playsound(loc, "polaroid", 75, 1, -3)
+		spawn(3 SECONDS)
+			triggered = FALSE
+		if(istimeagent(user))
+			var/datum/role/R = user.mind.GetRole(TIMEAGENT)
+			if(R)
+				var/datum/objective/target/locate/L = locate() in R.objectives.GetObjectives()
+				if(L)
+					L.check(view(target,2))
+
+/obj/item/weapon/gun/projectile/automatic/rewind
+	name = "rewind rifle"
+	desc = "Don't need to reload if you just rewind the bullets back into the gun."
+	icon_state = "xcomlasergun"
+	ammo_type = "/obj/item/ammo_casing/a12mm"
+	caliber = list(MM12 = 1)
+
+/obj/item/weapon/gun/projectile/automatic/rewind/send_to_past(var/duration)
+	var/mob/owner = loc
+	..()
+	spawn(duration)
+		owner.put_in_hands(src)
+
+/obj/item/weapon/gun/projectile/automatic/rewind/special_check(var/mob/M)
+	return istimeagent(M)
+
+/obj/item/weapon/gun/projectile/automatic/rewind/process_chambered()
+	attempt_past_send(rand(10,15) SECONDS)
```

Comment:
please test the gun and let me know if this works fine

Reply:
fire any number of times, the magazine resets to full after 10-15 seconds from the first shot elapses. works about as expected from testing.